src/vision/status_quo/niklaus_wants_to_use_ghostcell.md

```diff
@@ -0,0 +1,149 @@
+# Niklaus wants to use GhostCell-like cell borrowing with futures
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the
+brainstorming period. It is derived from real-life experiences of
+actual Rust users and is meant to reflect some of the challenges that
+Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to
+the FAQ, feel free to open a PR making edits (but keep in mind that,
+as they reflect peoples' experiences, status quo stories [cannot be
+wrong], only inaccurate). Alternatively, you may wish to [add your own
+status quo story][htvsq]!
+
+## The story
+
+Niklaus quite likes using statically-checked cell borrowing.  This is
+implemented in various ways in
+[GhostCell](http://plv.mpi-sws.org/rustbelt/ghostcell/) and
+[`qcell`](https://docs.rs/qcell/0.4.1/qcell/).  He would like to use
+statically-checked cell borrowing within futures, but there is no way
+to get the owner borrow through the `Future::poll` call.
+
+So he is forced to use `RefCell` instead and be very careful not to
+cause panics.  This seems like a step back.  It feels dangerous to use
+`RefCell` and to have to manually-verify that his cell borrows are
+panic-free.
+
+### The mechanism
+
+Statically-checked cell borrows work by having an owner held by the
+runtime and various instances of a cell held by things running on top
+of the runtime (these cells would typically be behind `Rc`
+references).  A mutable borrow on the owner is passed down the stack,
+which enables safe borrows on all the cells, since a mutable borrow on
+a cell is enabled by temporarily holding onto the mutable borrow of
+the owner.  This is all checked at compile-time.
+
+So the mutable borrow needs to be passed through the `poll` call, and
+it seems that requires support from the standard library.
+
+Right now a `&mut Context<'_>` is passed, and so within `Context`
+would be the ideal place to hold a borrow on the cell owner.  However
+as far as Niklaus can see there are difficulties with all the current
+implementations:
+
+- GhostCell (or qcell::LCell) is the best available solution, because
+  it doesn't have any restrictions on how many runtimes might be
+  running or whatever.  But Rust insists that the lifetimes `<'id>` on
+  methods are explicit, so it seems like that would force a change to
+  the signature of `poll`, which would break the ecosystem.
+
+- The other `qcell` types (QCell, TCell and TLCell) have various
+  restrictions or overheads which might make them unsuitable as a
+  general-purpose solution in the standard library.  However they do
+  have the positive feature of not requiring any change in the
+  signature of `poll`.  It looks like they could be added to `Context`
+  without breaking anything.
```

Comment:
I don't know what these types are -- it would be nice to give some examples, maybe? e.g., show how he is able to use them?